File: spec2code/java-spring/src/main/java/de/codecentric/javaspring/web/Problem.java
package de.codecentric.javaspring.web;

import org.springframework.http.HttpStatus;

public record Problem(String type, String title, int status, String detail) {
    public Problem(String title, int status, String detail) {
        this("about:blank", title, status, detail);
    }

    public Problem(HttpStatus status, Exception ex) {
        this(getType(status), status.getReasonPhrase(), status.value(), ex.getMessage());
    }

    public Problem(int status, Exception ex) {
        this(HttpStatus.valueOf(status).getReasonPhrase(), status, ex.getMessage());
    }

    public Problem(HttpStatus status, String detail) {
        this(status.getReasonPhrase(), status.value(), detail);
    }

    private static String getType(HttpStatus status) {
        return String.format("https://www.rfc-editor.org/rfc/rfc9110.html#name-%d-%s", status.value(),
                status.name().toLowerCase().replace("_", "-"));
    }
}

File: spec2code/kotlin-spring/src/main/kotlin/de/codecentric/kotlinspring/web/Problem.kt
package de.codecentric.kotlinspring.web

import de.codecentric.kotlinspring.api.model.ProblemDTO
import de.codecentric.kotlinspring.common.orElse
import org.springframework.http.HttpStatus

data class Problem(
    val type: String = "about:blank",
    val title: String,
    val status: Int,
    val detail: String? = null,
) {
    fun toDTO(): ProblemDTO = ProblemDTO(type, title, status, detail.orElse { "Something went wrong." })
}

fun createProblem(
    status: HttpStatus,
    ex: Exception,
): Problem = Problem(type = getType(status), title = status.reasonPhrase, status = status.value(), detail = ex.message)

fun createProblem(
    status: Int,
    ex: Exception,
): Problem = Problem(title = HttpStatus.valueOf(status).reasonPhrase, status = status, detail = ex.message)

fun createProblem(
    status: HttpStatus,
    detail: String,
): Problem = Problem(type = getType(status), title = status.reasonPhrase, status = status.value(), detail = detail)

fun getType(status: HttpStatus) =
    "https://www.rfc-editor.org/rfc/rfc9110.html#name-${status.value()}-${status.name.lowercase().replace("_", "-")}"